### DISCUSSION TITLE
Suggestion: if the behaviour features could be more user-friendly

### DISCUSSION THREAD

**Discussion by jackytrabbit** [nickcoutsos/keymap-editor#92] [Ideas]
This is a suggestion based on my personal preference. So I am not requesting anything, just want to give my thoughts on the beta versions of behaviour features. 
I think the combos and macros are very straightforward, but the hold-tap behaviour part is kinda confusing for newcomers.
I am using tap hold in Vial like this:
![image](https://s2.xptou.com/2023/03/27/642127c6bb5b5.png)
It's kinda hard for me to try to figure out how to set 'when tap and when hold' now in zmk
![image](https://files.catbox.moe/ena5o4.png)

ZMK and this keymap editor is great. ZMK is kinda complicate and keymap editor is better for users.
I totally understand currently keymap editor is more based on ZMK. I just thought this may be a way to try, less technical and more user friendly.

**Comment by nickcoutsos** [comment]
It's a question of balancing simplicity with flexibility.

I realize there is interest in an interface like the one you've shared, but implementing it without sacrificing flexibility is a significant challenge. I see a couple of options:

1. Abstract behaviors perfectly so that what you see in the editor is what you get. This is difficult because ZMK itself is evolving and I would have to understand both ZMK functionality and different use cases people want from it.
2. Provide separate "Basic" and "Advanced" interfaces which is reasonable but _significant_ work.

My strategy is to handle things as generically as possible so that in situations where I'm not caught up with the latest ZMK the user still has the escape hatch of editing their keymap directly and usually still load it into the keymap editor to manage the supported features.

Taking your example, assuming that a user could use each of those features simultaneously it would seem that the editor would have to generate:

1. a `hold-tap` with bindings based on the _"On hold"_ and _"On tap"_ selections
2. a second `hold-tap` with bindings based on the _"On tap + hold"_ and _"On double tap"_ selections
3. a `tap-dance` where the first bindings use the first and second defined behaviors

All of these behaviors can have their own configurations that this UI doesn't expose. What happens if a user has composed behaviors like this but their version accounts for triple-tapping? What about mod-morphs? What if the user has defined a custom hold-tap behavior in their keymap that they're referencing directly instead of the default `&mt`/`&lt`?

There _are_ answers to these questions, I don't want to dismiss the idea as being impossible, but there's no shortage of other questions I have to consider so I see it as being a little less urgent.

**Reply by jackytrabbit** [reply]
Yeah, I totally agree with that. The ZMK itself is whole new concept for users, and more complicated. It does't have to be like the others. I do find more and more interesting features in ZMK. I need more time to learn about it.
